nexus_constructor/component_tree_model.py

```diff
@@ -0,0 +1,284 @@
+        self.beginInsertRows(target_index, target_pos, target_pos)
+        transformation_link = LinkTransformation(transformation_list)
+        transformation_list.has_link = True
+        transformation_list.link = transformation_link
+        self.endInsertRows()
+
+    def add_component(self, new_component):
+        self.beginInsertRows(QModelIndex(), len(self.components), len(self.components))
+        self.components.append(new_component)
+        self.endInsertRows()
+
+    def remove_node(self, node):
+        if isinstance(node.internalPointer(), ComponentModel):
+            remove_index = self.components.index(node.internalPointer())
+            self.beginRemoveRows(QModelIndex(), remove_index, remove_index)
+            self.instrument.remove_component(node.internalPointer())
+            self.components.pop(remove_index)
+            self.endRemoveRows()
+        elif isinstance(node.internalPointer(), TransformationModel):
+            transformation_list = node.internalPointer().parent
+            transformation_list_index = self.parent(node)
+            remove_pos = transformation_list.index(node.internalPointer())
+            component = transformation_list.parent_component
+            self.beginRemoveRows(transformation_list_index, remove_pos, remove_pos)
+            component.remove_transformation(node.internalPointer())
+            transformation_list.pop(remove_pos)
+            self.endRemoveRows()
+        elif isinstance(node.internalPointer(), LinkTransformation):
+            transformation_list = node.internalPointer().parent
+            transformation_list_index = self.parent(node)
+            remove_pos = len(transformation_list)
+            self.beginRemoveRows(transformation_list_index, remove_pos, remove_pos)
+            transformation_list.has_link = False
+            transformation_list.link = None
+            self.endRemoveRows()
+            # Update depends on
+            if len(transformation_list) > 0:
+                parent_transform = transformation_list[len(transformation_list) - 1]
+                parent_transform.depends_on = None
+
+    def duplicate_node(self, node):
+        parent = node.internalPointer()
+        if isinstance(parent, ComponentModel):
+            new_name = get_duplication_name(parent.name, self.components)
+
+            self.add_component(
+                self.instrument.add_component(
+                    new_name, parent.nx_class, parent.description
+                )
+            )
+        elif isinstance(parent, TransformationModel):
+            raise NotImplementedError("Duplication of transformations not implemented")
+
+    def add_transformation(self, parent_index, type):
+        parentItem = parent_index.internalPointer()
+        transformation_list = None
+        parent_component = None
+        target_pos = 0
+        target_index = QModelIndex()
+        if isinstance(parentItem, ComponentModel):
```

Comment:
Would this be better solved by polymorphism?

Reply:
Maybe, I do not think this is a clear-cut case though. I want to make minimal changes to the implementations of `TransformationModel` and `ComponentModel`. Because the generation of instances of these classes is done through factory functions (`add_translation`/`add_rotation` and `add_component`) I can not easily subclass those either. The only alternative I can see is to have a model on top of the current one. I tried to do that first but it got messy. However, if there is some design (pattern) that I have missed feel free to suggest it.

Reply:
I've had a closer look and agree that polymorphism doesn't neatly solve this. A polymorhpic method to return `transformation_list`, `parent_component` and `target_pos` would be fine, but then you'd still have to return an enum or something to indicate how the `target_index` should be calculated. In the end I don't think it looks any better, so let's stick with how it is.